modules/core/src/test/scala/com.snowplowanalytics.snowplow.bigquery/processing/ProcessingSpec.scala

```diff
@@ -621,19 +632,117 @@ class ProcessingSpec extends Specification with CatsEffect {
 
+  def e13 = {
+    val processTime = Instant.parse("2023-10-24T10:00:42.123Z")
+
+    val eventID    = UUID.randomUUID()
+    val vCollector = "1.0.0"
+    val vEtl       = "2.0.0"
+    val source = goodOne(
+      optEventId    = Option(IO(eventID)),
+      optVCollector = Option(vCollector),
+      optVEtl       = Option(vEtl)
+    )
+
+    val io = runTest(inputEvents(count = 1, source), legacyColumnMode = true) { case (inputs, control) =>
+      for {
+        _ <- IO.sleep(processTime.toEpochMilli.millis)
+        _ <- Processing.stream(control.environment).compile.drain
+        state <- control.state.get
+      } yield {
+        val rows = state.writtenToBQ
+        (rows.size shouldEqual inputs.head.events.size) and
+          (rows.head should havePair("event_id" -> eventID.toString)) and
+          (rows.head should havePair("v_collector" -> vCollector)) and
+          (rows.head should havePair("v_etl" -> vEtl))
+      }
+    }
+    TestControl.executeEmbed(io)
+  }
+
+  def e14_base(legacyColumnMode: Boolean): IO[MatchResult[Vector[Action]]] = {
+    val collectorTstamp = Instant.parse("2023-10-24T10:00:00.000Z")
+    val processTime     = Instant.parse("2023-10-24T10:00:42.123Z")
+    val eventID         = UUID.randomUUID()
+    val vCollector      = "1.0.0"
+    val vEtl            = "2.0.0"
+
+    val data = json"""{ "myInteger": 100 }"""
+    val ue = UnstructEvent(
+      Some(SelfDescribingData(SchemaKey.fromUri("iglu:test_vendor/test_name/jsonschema/1-0-1").toOption.get, data))
+    )
+
+    val source = goodOne(
+      ue                 = ue,
+      optEventId         = Option(IO(eventID)),
+      optCollectorTstamp = Option(IO(collectorTstamp)),
+      optVCollector      = Option(vCollector),
+      optVEtl            = Option(vEtl)
+    )
+
+    val mocks = Mocks.default.copy(
+      addColumnsResponse = MockEnvironment.Response.Success(
+        FieldList.of(
+          BQField.of(
+            "unstruct_event_test_vendor_test_name_1",
+            StandardSQLTypeName.STRUCT,
+            FieldList.of(
+              BQField.of("my_string", StandardSQLTypeName.STRING),
+              BQField.of("my_integer", StandardSQLTypeName.INT64)
+            )
+          )
+        )
+      ),
+      descriptors = List(
+        AtomicDescriptor.withTestUnstruct100,
+        AtomicDescriptor.withTestUnstruct100,
+        AtomicDescriptor.withTestUnstruct101
+      )
+    )
+
+    val expectedColumnName =
+      if (legacyColumnMode) "unstruct_event_test_vendor_test_name_1_0_1"
+      else "unstruct_event_test_vendor_test_name_1"
+
+    val io = runTest(inputEvents(count = 1, source), mocks = mocks, legacyColumnMode = legacyColumnMode) { case (inputs, control) =>
+      for {
+        _ <- IO.sleep(processTime.toEpochMilli.millis)
+        _ <- Processing.stream(control.environment).compile.drain
+        state <- control.state.get
+      } yield state.actions should beEqualTo(
+        Vector(
```

Comment:
This is a nice improvement compared to when I reviewed it yesterday.

I think one tiny space for improvement... in this spec, currently this is the important line which covers the new feature that you added:

```scala
Action.AlterTableAddedColumns(Vector(expectedColumnName)),
```

But that's it.  In terms of testing the feature you added, currently you are just testing that it tried to alter the table with the expected column.

There is an opportunity to test more: You could also check `state.writtenToBQ`, and that will tell you whether it transformed the data in the expected way.

(Separately, I would like to amend some of the older tests to also check `state.writtenToBQ`, but that is beyond the scope of this PR).